### ISSUE TITLE
Why do PureComponents need props?

### ISSUE THREAD

**Issue by ZYinMD** [yannickcr/eslint-plugin-react#2186]
I'm using `"react/prefer-stateless-function": [1, { "ignorePureComponents": true }]`

When I have a PureComponent that doesn't have props, it shows a warning.

However, the component below WILL re-render if the render() of its parent is called
```
const Child = () => <h1>I re-renders when my parent does</h1>
```

You can see it with the React devtool in Chrome, turn on "highlight update"

If I change it to a PureComponent, it does not re-render:
```
class Child extends PureComponent {
  render() {
    return <h1>I don't</h1>;
  }
}
```
So to me, a PureComponent without props makes perfect sense.

**Comment by ZYinMD** [comment]
For a quick test yourself, use this code:
```
import React, { Component, PureComponent } from 'react';

export default class App extends Component {
  state = {};

  handleClick = () => {
    this.setState({
      foo: 'bar'
    });
  }

  render() {
    return (
    <div>
      <button onClick={this.handleClick}>re-render parent</button>
      <Child1 />
      <Child1 />
      <Child1 />
      <Child1 />
      <Child2 />
      <Child2 />
      <Child2 />
      <Child2 />
    </div>
    );
  }
}

const Child1 = () => <h1>I re-renders when my parent does</h1>;

class Child2 extends PureComponent {
  render() {
    return <h1>I don't</h1>;
  }
}

```

**Comment by ljharb** [comment]
Does the rule warn on Child2?

**Comment by ZYinMD** [comment]
Yes.

**Comment by ljharb** [comment]
With ignorePureComponents true, it’s not ignoring them, so that seems like a bug.

**Comment by ZYinMD** [comment]
It was intended, so you could say it's not a bug, but it enforces something that's not ideal:

https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-stateless-function.md#ignorepurecomponents

**Comment by ljharb** [comment]
oh. good point. I’d expect react not to rerender a component that gets no props or context, but it might not have chosen to apply that optimization.

**Comment by alexzherdev** [comment]
Perhaps on the off chance that it could depend on some global state?

**Comment by ljharb** [comment]
If it does, react explicitly doesn't guarantee a rerender - only props and state do that.

**Comment by pawelnvk** [comment]
Hopefully I understood you correctly. Please checkout: https://github.com/yannickcr/eslint-plugin-react/pull/2238.

**Comment by ljharb** [comment]
Fixed by #2238.
